### PR TITLE
Toolchain Modes

### PR DESCRIPTION
**Pull request by katre**
Tracking issue: https://github.com/bazelbuild/bazel/issues/3858

### CODE REVIEW

**Comment by katre** [comment]
Lead reviewer: @gregestren

**Comment by gregestren** [comment]
Review acknowledged.

**Comment by katre** [comment]
@laurentlb, it looks like @laszlocsomor is keeping the draft design doc in his own fork of the proposals repository. Is that an approach I should try as well?

**Comment by laurentlb** [comment]
From my point of view, both are okay (we're still trying to figure out the best workflow).

Adding the draft here means you need a pull request + approval for each change. The reviewer can point out typos, formatting or trivial issues. But GitHub pull requests are not well suited for the discussions. Since it was getting messy, @laszlocsomor moved his proposal to his GitHub, to keep all discussions in the mailing-list.

**Comment by laszlocsomor** [comment]
@katre : I'm happy to share my experience so far if you're interested.

File: designs/2018-07-23-toolchain-modes.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  - katre
+---
+
+# Background
+
+The current implementation of [toolchain
+resolution](https://docs.bazel.build/versions/master/toolchains.html#toolchain-resolution)
+is very simplistic: it assumes that there should only ever be one toolchain per
+target platform. However, that is not true: there are several reasons why there
+might be multiple valid toolchains for a given target platform:
```
Comment (gregestren) [inline designs/2018-07-23-toolchain-modes.md]: While I (think) I get what this is saying, I'm also a bit confused. Doesn't the current logic grant that different toolchains can match the platform but just happens to choose the first match?

If "assumes there should only be one toolchain per platform" means assume only one is *selected*, does that really change with this proposal? i.e. if you build `bazel build //foo --compilation_mode`, you can tweak the compilation mode for the build but that setting still applies to the entire target platform.

**Pull request closed without merging**